trin-core/src/portalnet/find/iterators/findcontent.rs

```diff
@@ -0,0 +1,715 @@
+                        let peer = peer.key().preimage().clone();
+                        return QueryState::Waiting(Some(peer));
+                    } else {
+                        return QueryState::WaitingAtCapacity;
+                    }
+                }
+
+                QueryPeerState::Waiting(timeout) => {
+                    if now >= *timeout {
+                        // Peers that don't respond within timeout are set to `Failed`.
+                        assert!(
+                            self.num_waiting > 0,
+                            "Query reached invalid number of waiting queries"
+                        );
+                        self.num_waiting -= 1;
+                        peer.set_state(QueryPeerState::Unresponsive);
+                    } else if at_capacity {
+                        // The query is still waiting for a result from a peer and is
+                        // at capacity w.r.t. the maximum number of peers being waited on.
+                        return QueryState::WaitingAtCapacity;
+                    } else {
+                        // The query is still waiting for a result from a peer and the
+                        // `result_counter` did not yet reach `num_results`. Therefore
+                        // the query is not yet done, regardless of already successful
+                        // queries to peers farther from the target.
+                        result_counter = None;
+                    }
+                }
+
+                QueryPeerState::Succeeded => {
+                    if let Some(ref mut count) = result_counter {
+                        *count += 1;
+                        // If `num_results` successful results have been delivered for the
+                        // closest peers, the query is done.
+                        if *count >= self.config.num_results {
+                            self.progress = QueryProgress::Finished;
+                            return QueryState::Finished;
+                        }
+                    }
+                }
+
+                QueryPeerState::Failed | QueryPeerState::Unresponsive => {
+                    // Skip over unresponsive or failed peers.
+                }
+            }
+        }
+
+        if self.num_waiting > 0 {
+            // The query is still waiting for results and not at capacity w.r.t.
+            // the allowed parallelism, but there are no new peers to contact
+            // at the moment.
+            QueryState::Waiting(None)
+        } else {
+            // The query is finished because all available peers have been contacted
+            // and the query is not waiting for any more results.
+            self.progress = QueryProgress::Finished;
+            QueryState::Finished
+        }
+    }
+
+    fn into_result(self) -> FindContentQueryResult<TNodeId> {
+        match self.content {
+            Some(content) => {
+                let closest_nodes = self
```

Comment:
This expression for getting `closest_nodes` and the one below in `None` seem to be almost identical - have you tried combining them into a shared method?

Reply:
They are similar but not identical. If the content is `Some(..)`, then we make sure to filter out the peer who returned the content. I could make this a shared method with an argument for a (possibly empty) set of nodes to filter out, but I did not think there was a huge benefit doing that.